src/database.ts

```diff
@@ -422,14 +456,17 @@ async function handleSetWorkspace(req: AuthenticatedRequest, res: express.Respon
 
     try {
         let updateResult: UpdateResult;
+        let outputWorkspace = {...(workspace as any), id: "", editable: true,  name: workspaceName};
         const existingWorkspace = await workspacesCollection.findOne({username: req.username, name: workspaceName});
         if (existingWorkspace) {
             updateResult = await workspacesCollection.updateOne({_id: existingWorkspace._id}, {$set: {workspace}}, {upsert: false});
+            outputWorkspace.id = existingWorkspace._id;
         } else {
             updateResult = await workspacesCollection.updateOne({username: req.username, name: workspaceName, workspace}, {$set: {workspace}}, {upsert: true});
```

Comment:
Why `updateOne` with `upsert: true` here instead of `insertOne`? (since you seem to have established that no workspace with the name already exists), or is there some worry about race conditions if the call is made twice in rapid succession for some reason or other?)

Is there some reason why the whole search for existing namespace and then, conditional on a match being found, different commands being run couldn't be condensed down into a single call to [`findOneAndUpdate`](https://mongodb.github.io/node-mongodb-native/5.7/classes/Collection.html#findOneAndUpdate) which might be slightly more efficient.  (Based on [this](https://mongodb.github.io/node-mongodb-native/4.0/interfaces/findoneandupdateoptions.html) it looks like `findOneAndUpdate` likely supports `upsert`).

Reply:
> Why updateOne with upsert: true here instead of insertOne? (since you seem to have established that no workspace with the name already exists), or is there some worry about race conditions if the call is made twice in rapid succession for some reason or other?)

I don't have a good answer for that. I will update it

> Is there some reason why the whole search for existing namespace and then, conditional on a match being found, different commands being run couldn't be condensed down into a single call to [findOneAndUpdate](https://mongodb.github.io/node-mongodb-native/5.7/classes/Collection.html#findOneAndUpdate) which might be slightly more efficient. (Based on [this](https://mongodb.github.io/node-mongodb-native/4.0/interfaces/findoneandupdateoptions.html) it looks like findOneAndUpdate likely supports upsert).

I will look into this too. Didn't realise that also supported upsert.

Reply:
changed to `findOneAndUpdate` :+1: